bookshelf/test/functional/controllersTest.php

```diff
@@ -0,0 +1,205 @@
+     */
+    public function createApplication()
+    {
+        $app = require __DIR__ . '/../../src/app.php';
+        require __DIR__ . '/../../src/controllers.php';
+        $app['bookshelf.model'] = new CloudsqlModel();
+        // Set a tiny page size so it's easy to test paging.
+        $app['bookshelf.page_size'] =  1;
+        $this->storage = $app['bookshelf.storage'] = new FakeImageStorage();
+
+        return $app;
+    }
+
+    public function testRoot()
+    {
+        $client = $this->createClient();
+        $client->request('GET', '/');
+
+        $this->assertTrue($client->getResponse()->isRedirect());
+    }
+
+    public function testPaging()
+    {
+        $client = $this->createClient();
+        $client->followRedirects();
+        $crawler = $client->request('GET', '/books/');
+
+        $editLink = $crawler
+            ->filter('a:contains("Add")') // find all links with the text "Add"
+            ->link();
+
+        $crawler = $client->click($editLink);
+
+        // Fill the form and submit it, twice.
+        $submitButton = $crawler->selectButton('submit');
+        $form = $submitButton->form();
+
+        $photo = new UploadedFile(
+            __DIR__.'/../lib/CatHat.jpg',
+            'CatHat.jpg',
+            'image/jpg',
+            filesize(__DIR__.'/../lib/CatHat.jpg')
+        );
+        $crawler = $client->submit($form, array(
+            'title' => 'The Cat in the Hat',
+            'author' => 'Dr. Suess',
+            'publishedDate' => '1957-01-01',
+            'image' => $photo
+        ));
+        $this->assertEquals(
+            'img1',
+            $crawler->filter('.book-image')->attr('src')
+        );
+
+        // Capture the delete button.
+        $deleteCatHat = $crawler->selectButton('submit');
+
+        $crawler = $client->submit($form, array(
+            'title' => 'Treasure Island',
+            'author' => 'Robert Louis Stevenson',
+            'publishedDate' => '1883-01-01'
+        ));
+        $deleteTreasureIsland = $crawler->selectButton('submit');
+
+        try {
+            // Now go through the pages one by one and confirm we saw the books
+            // we just added.
+            $foundTreasureIsland = false;
+            $foundCatHat = false;
+            $crawler = $client->request('GET', '/');
+            while (true) {
+                $foundCatHat = $foundCatHat ||
+                    $crawler->filter('h4:contains("The Cat in the Hat")');
+                $foundTreasureIsland = $foundTreasureIsland ||
+                    $crawler->filter('h4:contains("Treasure Island")');
+                $more = $crawler->filter('a:contains("More")');
+                if (count($more)) {
+                    $crawler = $client->click($more->link());
+                } else {
+                    break;
+                }
+
+            }
+            $this->assertTrue($foundTreasureIsland);
+            $this->assertTrue($foundCatHat);
```

Comment:
It seems like we should get rid of the `try` here, but if it's necessary, we should add a `$this->fail(..)` in the `catch`, otherwise the test will succeed even if an exception is thrown and the images are not found.